### PR TITLE
KOY-3129: Was added availability to work directly with buffer without…

### PR DESCRIPTION
**Pull request by develstyle**
… json parse/stringify

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,11 +21,14 @@
...
+    "build": "tsc --noEmit false && cp ./src/types.d.ts ./lib/types.d.ts",
     "test": "npm run tslint && mocha --compilers ts:ts-node/register",
     "test:docker": "docker-compose -f docker-compose.test.yml up --build",
     "tslint": "tslint -c tslint.json -p tsconfig.json",
-    "watch": "tsc -w",
+    "tsc": "tsc -p .",
+    "tsc:withEmit": "tsc --noEmit false -p .",
+    "tsc:clean": "find ./src ./test -type f -name '*.js' -delete && find ./src ./test -type f -name '*.js.map' -delete",
+    "watch2": "tsc -w",
     "precommit": "npm run test"
```
Comment (maozedong) [inline package.json]: add something like prepublish script please, where tests and build will be executed
https://docs.npmjs.com/misc/scripts#description

File: package.json [hunk 2]
```diff
@@ -21,11 +21,14 @@
...
-    "build": "tsc && cp ./src/types.d.ts ./lib/types.d.ts",
+    "build": "tsc --noEmit false && cp ./src/types.d.ts ./lib/types.d.ts",
     "test": "npm run tslint && mocha --compilers ts:ts-node/register",
     "test:docker": "docker-compose -f docker-compose.test.yml up --build",
     "tslint": "tslint -c tslint.json -p tsconfig.json",
-    "watch": "tsc -w",
+    "tsc": "tsc -p .",
+    "tsc:withEmit": "tsc --noEmit false -p .",
+    "tsc:clean": "find ./src ./test -type f -name '*.js' -delete && find ./src ./test -type f -name '*.js.map' -delete",
+    "watch2": "tsc -w",
```
Comment (maozedong) [inline package.json]: why watch2?

File: src/rabbitMQWorker.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ export class RabbitMQWorker implements BusWorker {
   private _channel: amqplib.Channel
   private url: string
   private responseEmitter: NodeJS.EventEmitter
+  private _json: boolean = true
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: let's not use default json option in worker, it doesn't make much sense

File: src/rabbitMQWorker.ts [hunk 4]
```diff
@@ -55,11 +56,12 @@ export class RabbitMQWorker implements BusWorker {
     return this._channel.publish(exchange, key, content, options)
   }
 
-  async subscribe (queue, eventEmitter: NodeJS.EventEmitter, noAck) {
+  async subscribe (queue, eventEmitter: NodeJS.EventEmitter, noAck, json?: boolean) {
+    this._json = json
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: setting json here to global one will break other subscribers. don't do that

File: src/rabbitMQWorker.ts [hunk 5]
```diff
@@ -55,11 +56,12 @@ export class RabbitMQWorker implements BusWorker {
...
     const options = {noAck}
     const {consumerTag} = await this._channel.consume(queue, (message) => {
       try {
-        const content = RabbitMQWorker.getMessageContent(message)
+        const content = this._json ? RabbitMQWorker.getMessageContent(message) : message
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: use json argument

File: src/rabbitMQWorker.ts [hunk 6]
```diff
@@ -81,7 +83,8 @@ export class RabbitMQWorker implements BusWorker {
   }
 
   request (options) {
-    const {key, exchange, timeout, route, message} = options
+    const {key, exchange, timeout, route, message, json} = options
+    this._json = json
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: same thing about global _json

File: src/rabbitMQWorker.ts [hunk 7]
```diff
@@ -94,11 +97,12 @@ export class RabbitMQWorker implements BusWorker {
 
       this.responseEmitter.once(correlationId, (msg) => {
         clearTimeout(timeoutId)
-        const content = RabbitMQWorker.getMessageContent(msg)
+        const content = this._json ? RabbitMQWorker.getMessageContent(msg) : msg
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: don't use this._json

File: src/rabbitMQWorker.ts [hunk 8]
```diff
@@ -94,11 +97,12 @@ export class RabbitMQWorker implements BusWorker {
...
         return resolve({msg, content})
       })
 
-      this._channel.publish(exchange, key, Buffer.from(JSON.stringify(message)), {
+      const data = this._json ? Buffer.from(JSON.stringify(message)) : message
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: don't use this._json

File: src/rabbitMQWorker.ts [hunk 9]
```diff
@@ -108,7 +112,8 @@ export class RabbitMQWorker implements BusWorker {
 
   async respond (res, msg) {
     const {replyTo, correlationId} = msg.properties
-    return this._channel.publish('', replyTo, Buffer.from(JSON.stringify(res)), {correlationId})
+    const data = this._json ? Buffer.from(JSON.stringify(res)) : res.content
```
Comment (maozedong) [inline src/rabbitMQWorker.ts]: don't use this._json

File: test/requester.ts [hunk 10]
```diff
@@ -69,4 +69,22 @@ describe('requester', function () {
 
   })
```
Comment (maozedong) [inline test/requester.ts]: add test to check whether subscribers and/or requesters with different 'json' option can exist simultaneously

File: src/rabbitMQWorker.ts (hunk 4 above)
Comment (develstyle) [reply src/rabbitMQWorker.ts]: Need to discuss it.

File: src/rabbitMQWorker.ts (hunk 5 above)
Comment (develstyle) [reply src/rabbitMQWorker.ts]: Was refactored with json as argument.

File: src/rabbitMQWorker.ts (hunk 6 above)
Comment (develstyle) [reply src/rabbitMQWorker.ts]: Was refactored.

File: src/rabbitMQWorker.ts (hunk 7 above)
Comment (develstyle) [reply src/rabbitMQWorker.ts]: Was refactored.

File: src/rabbitMQWorker.ts (hunk 9 above)
Comment (develstyle) [reply src/rabbitMQWorker.ts]: How do you propose to put json var value inside this method?

File: test/requester.ts (hunk 10 above)
Comment (develstyle) [reply test/requester.ts]: We have such tests.

File: src/rabbitMQWorker.ts [hunk 11]
```diff
@@ -55,11 +55,11 @@ export class RabbitMQWorker implements BusWorker {
     return this._channel.publish(exchange, key, content, options)
   }
 
-  async subscribe (queue, eventEmitter: NodeJS.EventEmitter, noAck) {
+  async subscribe (queue, eventEmitter: NodeJS.EventEmitter, noAck, json?: boolean) {
     const options = {noAck}
     const {consumerTag} = await this._channel.consume(queue, (message) => {
       try {
-        const content = RabbitMQWorker.getMessageContent(message)
+        const content = json ? RabbitMQWorker.getMessageContent(message) : message
```
[the remark this answers is not in the document]
Comment (develstyle) [reply src/rabbitMQWorker.ts]: Was refactored with message.content

File: src/rabbitMQWorker.ts [hunk 12]
```diff
@@ -94,21 +94,23 @@ export class RabbitMQWorker implements BusWorker {
 
       this.responseEmitter.once(correlationId, (msg) => {
         clearTimeout(timeoutId)
-        const content = RabbitMQWorker.getMessageContent(msg)
+        const content = json ? RabbitMQWorker.getMessageContent(msg) : msg
```
[the remark this answers is not in the document]
Comment (develstyle) [reply src/rabbitMQWorker.ts]: Was refactored with message.content

**Comment by develstyle** [comment]
about tests:
subscriber.ts -> subscriber should receive published msg as buffer
requester.ts -> responder should respond with request msg as buffer
responder -> responder should respond with request msg as buffer

**Pull request merged**